cumulus/pallets/weight-reclaim/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+		len: usize,
+	) -> Result<Self::Pre, TransactionValidityError> {
+		let (proof_size, inner_val) = val;
+		self.0.prepare(inner_val, origin, call, info, len).map(|pre| (proof_size, pre))
+	}
+
+	fn post_dispatch_details(
+		pre: Self::Pre,
+		info: &DispatchInfoOf<T::RuntimeCall>,
+		post_info: &PostDispatchInfoOf<T::RuntimeCall>,
+		len: usize,
+		result: &DispatchResult,
+	) -> Result<Weight, TransactionValidityError> {
+		let (pre_dispatch_proof_size, inner_pre) = pre;
+
+		let mut post_info_with_inner = *post_info;
+		S::post_dispatch(inner_pre, info, &mut post_info_with_inner, len, result)?;
+
+		let inner_refund = if let (Some(before_weight), Some(after_weight)) =
+			(post_info.actual_weight, post_info_with_inner.actual_weight)
+		{
+			before_weight.saturating_sub(after_weight)
+		} else {
+			Weight::zero()
+		};
+
+		let Some(pre_dispatch_proof_size) = pre_dispatch_proof_size else {
+			// We have no proof size information, there is nothing we can do.
+			return Ok(inner_refund);
+		};
+
+		let Some(post_dispatch_proof_size) = get_proof_size() else {
+			log::debug!(
+				target: LOG_TARGET,
+				"Proof recording enabled during prepare, now disabled. This should not happen."
+			);
+			return Ok(inner_refund)
+		};
+
+		// The consumed proof size as measured by the host.
+		let measured_proof_size = post_dispatch_proof_size.saturating_sub(pre_dispatch_proof_size);
+
+		// The consumed weight as benchamrked.
+		// NOTE: `calc_actual_weight` will take the minimum of `post_info` and `info` weights.
+		// This means any underestimation of compute time in the pre dispatch info will not be
+		// taken into account.
+		let benchmarked_weight = post_info_with_inner.calc_actual_weight(info);
+
+		let benchmarked_proof_size = benchmarked_weight.proof_size();
+		if benchmarked_proof_size < measured_proof_size {
+			log::error!(
+				target: LOG_TARGET,
+				"Benchmarked storage weight smaller than consumed storage weight. \
+				benchmarked: {benchmarked_proof_size} consumed: {measured_proof_size}"
+			);
+		} else {
+			log::trace!(
+				target: LOG_TARGET,
+				"Reclaiming storage weight. benchmarked: {benchmarked_proof_size},
+				consumed: {measured_proof_size}"
+			);
+		}
+
+		let accurate_weight = benchmarked_weight.set_proof_size(measured_proof_size);
+
+		frame_system::BlockWeight::<T>::mutate(|current_weight| {
+			let already_reclaimed = frame_system::ExtrinsicWeightReclaimed::<T>::get();
+			current_weight.accrue(already_reclaimed, info.class);
+			current_weight.reduce(info.total_weight(), info.class);
+			current_weight.accrue(accurate_weight, info.class);
+
+			// The saturation will happen if the pre dispatch weight is underestimating the proof
+			// size.
+			// In this case the extrinsic proof size weight reclaimed is 0.
+			let accurate_unspent = info.total_weight().saturating_sub(accurate_weight);
+			frame_system::ExtrinsicWeightReclaimed::<T>::put(accurate_unspent);
+		});
+
+		Ok(inner_refund)
```

Comment:
Shouldn't we refund the overestimated proof sizes of the inner extensions? I read the function a bunch of times but I don't think we do, the proof size overestimation we'd need to refund would be something like `post_info_with_inner.calc_actual_weight(info).proof_size().saturating_sub(measured_proof_size)` and we'd need to add this to `inner_refund`. Am I missing anything?

Reply:
Indeed I didn't bother returning accurate unspent amount.

If we include the unspent amount of all the dispatch (call, and inner transaction extension), then the unspent we can return is `info.total_weight().saturating_sub(accurate_weight)`.

I will return this and also ensure it is tested

Reply:
done in https://github.com/paritytech/polkadot-sdk/pull/6140/commits/597fde45aed898df8880eaaed0dedcee0255ef56